src/util/getGenotype.js

```diff
@@ -20,6 +20,13 @@ export const encodeColorByGenotype = ({ gene, positions }) => {
   return `gt-${gene}_${positions.join(",")}`;
 };
 
+export const decodePositions = (positions, geneLength = 'Infinity') => {
+  return positions
+    .split(",")
+    .map((x) => parseInt(x, 10))
+    .filter((x) => x > 0 && x <= Math.floor(geneLength));
+};
+
 /* Examples:
  *   gt-nuc_142       → { gene: "nuc", aa: false, positions: [142] }
  *   gt-nuc_142,144   → { gene: "nuc", aa: false, positions: [142,144] }
```

Comment:
> These weren't flagged as errors previously, however the previous commit swapped the rule "no-use-before-define" to "@typescript/no-use-before-define".
> […]
> I don't know enough about specifications to know why this is considered an error but still works in practice, however it's not much effort to reorganise the code in these 2 files.

I find this sort of reorganization actively harmful to the ability to read the code in a file. Reading top to bottom, I want to read the higher level functions first, which call the lower level functions that I'll read next. The opposite is like putting the cart before the horse. `decodePositions` is out of context and makes no sense until I see it called in `decodeColorByGenotype`.

This is why in Auspice 6f22d9f87e3e8e16662a1717b527d2351fe44ff1 disabled `no-use-before-define` for functions, based on [my change to nextstrain.org's config](https://github.com/nextstrain/nextstrain.org/commit/fc4aee33db580dbe0a76529f19137e24b8ce31c4).

Reply:
There's a few things at play here.

Firstly, I didn't notice the existing eslint overrule for that line [1]:

https://github.com/nextstrain/auspice/blob/b58f14889924ca887587d5a65d623b97d0ed1b32/src/util/getGenotype.js#L45

With this PR this overrule needs to change to `// eslint-disable-line @typescript-eslint/no-use-before-define`, which I'll do as part of this PR. (This keeps the code as intended, which is appropriate for this PR.)


Secondly, I don't know why the following change was made (by me, but over a year ago):

```diff
-  no-use-before-define: ["error", { "functions": false }]
+  no-use-before-define: "off"  # note you must disable the base rule as it can report incorrect errors
+  "@typescript-eslint/no-use-before-define": ["error", { "functions": false }]
```

The intention is to allow function (not fat-arrow) hoisting as per https://github.com/nextstrain/auspice/commit/6f22d9f87e3e8e16662a1717b527d2351fe44ff1 (rule docs [here](https://eslint.org/docs/latest/rules/no-use-before-define#functions)). Linting behavior under `no-use-before-define": ["error", { "functions": false }]`:

```
# Not valid JS, and has linting errors
const foo = () => bar();
foo();
const bar = () => {console.log("42");};

# Valid JS, and no linting errors
const foo = () => bar();
foo();
function bar() {console.log("42");}

# Valid JS, but linting fails because of the introduction
# of a temporal dead zone. Allowable by using
# ["error", { "functions": false, "variables": false}]
const foo = () => bar();
const bar = () => {console.log("42");};
foo();
```

The third example is why we get the errors in `choose-layout.js`, however why they were flagged up now and not before I don't understand 🤷 . Note that while setting `"variables": false` allows the existing code in `choose-layout.js` it also allows the first toy example above to pass linting, despite it throwing a `ReferenceError` when run. 

[1] You wrote that override a ~year before we allowed function hoisting in https://github.com/nextstrain/auspice/commit/6f22d9f87e3e8e16662a1717b527d2351fe44ff, but it's not a function declaration so it wouldn't be affected by that rule anyways.

Reply:
Compared ASTs for this toy example:

```js
class Foo extends React.Component {
  renderX() {
    return (<Bar />);
  }
}

const Bar = ({ }) => (<>bar</>);
```

using https://astexplorer.net with `@babel/eslint-parser` and `@typescript-eslint/parser`. Sorting the JSON output and [diffing it](https://gist.github.com/tsibley/06761e22bff4fdc812a348b1871e3dd7) produces nothing obvious. It's very similar. So maybe Typescript's `no-use-before-define` rule supports `JSXIdentifier` nodes and the standard rule just… doesn't?

OTOH, signs still point to a parsing difference when I interrogate the example above with `npx eslint --ext .js,.ts,.jsx,.tsx --parser @babel/eslint-parser` vs. `… --parser @typescript-eslint/parser` and also try switching between the two rules.

parser / rule | `no-use-before-define` | `@typescript-eslint/no-use-before-define`
--- | --- | ---
`@babel/eslint-parser` | ok | ok
`@typescript-eslint/parser` | ok | errors

Reply:
Nice work going down that rabbit hole!

As part of PR #1665, https://github.com/nextstrain/auspice/commit/9fc42f01ee6609631a9d864da7e8c4da5cf5e583 turned `decodePositions` into a function declaration which is hoisted:

```diff
- export const decodePositions = (positions, geneLength = 'Infinity') => {
+ export function decodePositions(positions, geneLength = 'Infinity') {
```
This is left unchanged by this PR.

The changes to `choose-layout.js` still remain in this PR, due to the change in interpretation between `@babel/eslint-parser` and `@typescript-eslint/parser` (as per Tom's message above).

I think this thread can be considered resolved.

The other cha